mobile_verifier/src/subscriber_mapping_activity.rs

```diff
@@ -0,0 +1,271 @@
+                let status = self.verify_activity(&sma).await?;
+                Ok((sma, status))
+            })
+            .and_then(|(sma, status)| self.write_verified_report(sma, status))
+            .try_filter_map(|(sma, status)| is_valid(sma, status));
+
+        db::save(&mut transaction, activity_stream).await?;
+        self.verified_sink.commit().await?;
+        transaction.commit().await?;
+        Ok(())
+    }
+
+    async fn write_verified_report(
+        &self,
+        mut activity: SubscriberMappingActivity,
+        status: SubscriberReportVerificationStatus,
+    ) -> anyhow::Result<(
+        SubscriberMappingActivity,
+        SubscriberReportVerificationStatus,
+    )> {
+        let verified_proto = VerifiedSubscriberMappingActivityReportV1 {
+            report: activity.take_original(),
+            status: status as i32,
+            timestamp: Utc::now().encode_timestamp_millis(),
+        };
+
+        self.verified_sink
+            .write(verified_proto, &[("status", status.as_str_name())])
+            .await?;
+
+        Ok((activity, status))
+    }
+
+    async fn verify_activity(
+        &self,
+        activity: &SubscriberMappingActivity,
+    ) -> anyhow::Result<SubscriberReportVerificationStatus> {
+        if !self
+            .verify_known_carrier_key(&activity.carrier_pub_key)
+            .await?
+        {
+            return Ok(SubscriberReportVerificationStatus::InvalidCarrierKey);
+        };
+        if !self.verify_subscriber_id(&activity.subscriber_id).await? {
+            return Ok(SubscriberReportVerificationStatus::InvalidSubscriberId);
+        };
+        Ok(SubscriberReportVerificationStatus::Valid)
+    }
+
+    async fn verify_known_carrier_key(&self, public_key: &PublicKeyBinary) -> anyhow::Result<bool> {
+        self.authorization_verifier
+            .verify_authorized_key(public_key, NetworkKeyRole::MobileCarrier)
+            .await
+            .map_err(anyhow::Error::from)
+    }
+
+    async fn verify_subscriber_id(&self, subscriber_id: &[u8]) -> anyhow::Result<bool> {
+        self.entity_verifier
+            .verify_rewardable_entity(subscriber_id)
+            .await
+            .map_err(anyhow::Error::from)
+    }
+}
+
+async fn is_valid(
+    activity: SubscriberMappingActivity,
+    status: SubscriberReportVerificationStatus,
+) -> anyhow::Result<Option<SubscriberMappingActivity>> {
+    if status == SubscriberReportVerificationStatus::Valid {
+        Ok(Some(activity))
+    } else {
+        Ok(None)
+    }
+}
+
+impl<AV, EV> ManagedTask for SubscriberMappingActivityDaemon<AV, EV>
+where
+    AV: AuthorizationVerifier,
+    EV: EntityVerifier + Send + Sync + 'static,
+{
+    fn start_task(
+        self: Box<Self>,
+        shutdown: triggered::Listener,
+    ) -> futures::future::LocalBoxFuture<'static, anyhow::Result<()>> {
+        Box::pin(self.run(shutdown))
+        // let handle = tokio::spawn(async move { self.run(shutdown).await });
```

Comment:
@michaeldjeffrey I am hoping to actually spawn this, but when i do I get a weird lifetime issue that I haven't figured out yet, maybe you got some ideas?

Reply:
Addressed the lifetime issues here c68a9fb, not sure if it is the best way, but it seems to work